config.org
#+title: Example Rational Emacs User Configuration File
#+PROPERTY: header-args:emacs-lisp :tangle ./config.el :mkdirp yes

Rational Emacs supports user customization through a =config.org= file similar to
this one.

To get you started, copy both example files to your Rational Emacs configuration
directory, e.g.:
- copy =examples/example-config.el= to =~/.rational-emacs/config.el=
- copy =examples/example-config.org= to =~/.rational-emacs/config.org=
- restart Emacs

After that, whenever you edit =config.org= in Emacs and save it, it updates
=config.el= in the same directory.

See the [[https://orgmode.org/manual/Extracting-Source-Code.html][Org Manual]] for details and alternatives.

#+begin_src emacs-lisp
  ;;; config.el -- Example Rational Emacs user customization file -*- lexical-binding: t; -*-
  ;; This file is generated from config.org. If you want to edit the
  ;; configuration, DO NOT edit config.el, edit config.org, instead.

  ;; Tangle the code blocks to config.el on save.
  (defun org-babel-tangle-config ()
    (when (string-equal (buffer-file-name)
                        (expand-file-name "config.org" rational-config-path))
      ;; Dynamic scoping to the rescue
      (let ((org-confirm-babel-evaluate nil))
        (org-babel-tangle))))

  (add-hook 'org-mode-hook
            (lambda ()
              (add-hook 'after-save-hook #'org-babel-tangle-config)))
#+end_src

In your configuration you can set any Emacs configuration variable, face
attributes, themes, etc as you normally would.
See the README.org file in this repository for additional information.

* Rational Modules

At the moment, Rational Emacs offers the following modules. Comment out
everything you don't want to use.
At the very least, you should decide whether or not you want to use [[https://github.com/emacs-evil/evil][evil-mode]],
as it will greatly change how you interact with Emacs. So, if you prefer
Vim-style keybindings over vanilla Emacs keybindings remove the comment
in the line about =rational-evil= below.
#+begin_src emacs-lisp
  (require 'rational-defaults)    ; Sensible default settings for Emacs
  (require 'rational-use-package) ; Configuration for `use-package`
  (require 'rational-updates)     ; Tools to upgrade Rational Emacs
  (require 'rational-completion)  ; selection framework based on `vertico`
  (require 'rational-ui)          ; Better UI experience (modeline etc.)
  (require 'rational-windows)     ; Window management configuration
  (require 'rational-editing)     ; Whitspace trimming, auto parens etc.
  ;(require 'rational-evil)        ; An `evil-mode` configuration
  (require 'rational-org)         ; org-appear, clickable hyperlinks etc.
  (require 'rational-project)     ; built-in alternative to projectile
  (require 'rational-speedbar)    ; built-in file-tree
  (require 'rational-screencast)  ; show current command and binding in modeline
  (require 'rational-compile)     ; automatically compile some emacs lisp files
#+end_src
* Font and theme settings
#+begin_src emacs-lisp
  (custom-set-variables
     '(rational-ui-default-font
       '(:font "JetBrains Mono" :weight light :height 185)))

  (rational-package-install-package 'doom-themes)
  (load-theme 'doom-city-lights t)
#+end_src
* Doom Modeline
To make the information go somewhere visible

#+BEGIN_SRC emacs-lisp
 (use-package doom-modeline
    :ensure t
    :init (doom-modeline-mode 1)
    :custom ((doom-modeline-height 35)
             (doom-modeline-bar-width 10)
             (doom-modeline-buffer-file-name-style 'auto)
             (doom-modeline-icon (display-graphic-p))
             (doom-modeline-major-mode-icon t)
             (doom-modeline-major-mode-color-icon t)
             (doom-modeline-enable-word-count t)
             (doom-modeline-modal-icon t)
             (doom-modeline-mu4e nil)
             (mu4e-alert-enable-mode-line-display)))

  (display-battery-mode 1)

  ;; Turn on `display-time-mode' if you don't use an external bar.
  (setq display-time-default-load-average nil)

  (display-time-mode t)

  (setq display-time-24hr-format 1)

  ;;(display-time-mode nil)
  ;;(setq display-time-24hr-format nil)
#+END_SRC

* Let's get encrypted
#+begin_src emacs-lisp
(setenv "GPG_AGENT_INFO" nil)  ;; use emacs pinentry
    (setq auth-source-debug t)

    (setq epg-gpg-program "gpg2")  ;; not necessary
    (require 'epa-file)
    (epa-file-enable)
;;(setq epa-pinentry-mode 'loopback)
(setq epg-pinentry-mode 'loopback)

    (require 'org-crypt)
    (org-crypt-use-before-save-magic)

(defun efs/lookup-password (&rest keys)
  (let ((result (apply #'auth-source-search keys)))
    (if result
	(funcall (plist-get (car result) :secret))
      nil)))
#+end_src
* Ivy,Swiper,Avy

#+BEGIN_SRC emacs-lisp

  (use-package counsel
    :ensure t)
  
  (use-package ivy
    :ensure t
    :diminish (ivy-mode)
    :bind (("C-x b" . ivy-switch-buffer))
    :config
    (ivy-mode 1)
    (setq ivy-use-virtual-buffers t)
    (setq ivy-display-style 'fancy)
    (setq enable-recursive-minibuffers t)
  )
    (use-package swiper
      :ensure try
      :bind (("C-s" . swiper)
	     ("C-r" . swiper)
	     ("C-c C-r" . ivy-resume)
	     ("M-x" . counsel-M-x)
	     ("C-x C-f" . counsel-find-file))
      :config
      (progn
	(ivy-mode 1)
	(setq ivy-use-virtual-buffers t)
	(setq ivy-display-style 'fancy)
	(define-key read-expression-map (kbd "C-r") 'counsel-expression-history)
	))

  (use-package avy
    :ensure t
    :bind ("M-s" . avy-goto-char))
#+END_SRC
* ORG MODE

#+BEGIN_SRC emacs-lisp
  (defun efs/org-mode-setup ()
    (org-indent-mode)
    (variable-pitch-mode 1)
    (visual-line-mode 1))

    (setq org-ellipsis " ▾")

  (use-package org-bullets
    :hook (org-mode . org-bullets-mode)
    :custom
    (org-bullets-bullet-list '("◉" "○" "●" "○" "●" "○" "●")))

  (defun efs/org-mode-visual-fill ()
    (setq visual-fill-column-width 80
	  visual-fill-column-center-text t)
    (visual-fill-column-mode 1))

  (use-package visual-fill-column
    :hook (org-mode . efs/org-mode-visual-fill))

   (setq org-latex-pdf-process
	    '("latexmk -pdflatex='pdflatex -interaction nonstopmode' -pdf -bibtex -f %f"))

  (require 'org-tempo)


  (unless (boundp 'org-latex-classes)
    (setq org-latex-classes nil))

      (add-to-list 'org-latex-classes
		   '("ethz"
		     "\\documentclass[a4paper,11pt,titlepage]{memoir}
      \\usepackage[utf8]{inputenc}
      \\usepackage[T1]{fontenc}
      \\usepackage{fixltx2e}
      \\usepackage{graphicx}
      \\usepackage{longtable}
      \\usepackage{float}
      \\usepackage{wrapfig}
      \\usepackage{rotating}
      \\usepackage[normalem]{ulem}
      \\usepackage{amsmath}
      \\usepackage{textcomp}
      \\usepackage{marvosym}
      \\usepackage{wasysym}
      \\usepackage{amssymb}
      \\usepackage{hyperref}
      \\usepackage{mathpazo}
      \\usepackage{color}
      \\usepackage{enumerate}
      \\definecolor{bg}{rgb}{0.95,0.95,0.95}
      \\tolerance=1000
	    [NO-DEFAULT-PACKAGES]
	    [PACKAGES]
	    [EXTRA]
      \\linespread{1.1}
      \\hypersetup{pdfborder=0 0 0}"
		     ("\\chapter{%s}" . "\\chapter*{%s}")
		     ("\\section{%s}" . "\\section*{%s}")
		     ("\\subsection{%s}" . "\\subsection*{%s}")
		     ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
		     ("\\paragraph{%s}" . "\\paragraph*{%s}")
		     ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))


      (add-to-list 'org-latex-classes
		   '("article"
		     "\\documentclass[11pt,a4paper]{article}
      \\usepackage[utf8]{inputenc}
      \\usepackage[T1]{fontenc}
      \\usepackage{fixltx2e}
      \\usepackage{graphicx}
      \\usepackage{longtable}
      \\usepackage{float}
      \\usepackage{wrapfig}
      \\usepackage{rotating}
      \\usepackage[normalem]{ulem}
      \\usepackage{amsmath}
      \\usepackage{textcomp}
      \\usepackage{marvosym}
      \\usepackage{wasysym}
      \\usepackage{amssymb}
      \\usepackage{hyperref}
      \\usepackage{mathpazo}
      \\usepackage{color}
      \\usepackage{enumerate}
      \\definecolor{bg}{rgb}{0.95,0.95,0.95}
      \\tolerance=1000
	    [NO-DEFAULT-PACKAGES]
	    [PACKAGES]
	    [EXTRA]
      \\linespread{1.1}
      \\hypersetup{pdfborder=0 0 0}"
		     ("\\section{%s}" . "\\section*{%s}")
		     ("\\subsection{%s}" . "\\subsection*{%s}")
		     ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
		     ("\\paragraph{%s}" . "\\paragraph*{%s}")))


      (add-to-list 'org-latex-classes '("ebook"
					"\\documentclass[11pt, oneside]{memoir}
      \\setstocksize{9in}{6in}
      \\settrimmedsize{\\stockheight}{\\stockwidth}{*}
      \\setlrmarginsandblock{2cm}{2cm}{*} % Left and right margin
      \\setulmarginsandblock{2cm}{2cm}{*} % Upper and lower margin
      \\checkandfixthelayout
      % Much more laTeX code omitted
      "
					("\\chapter{%s}" . "\\chapter*{%s}")
					("\\section{%s}" . "\\section*{%s}")
					("\\subsection{%s}" .
  "\\subsection*{%s}")))
#+END_SRC
* Mu4e email configuration
#+begin_src emacs-lisp
(use-package mu4e
  :ensure nil
  :config

  ;;this is set to 't' to avoid mail syncing issues when using mbsync
  (setq mu4e-change-filenames-when-moving t)

  ;; refresh mail using isync every 10 minutes
  (setq mu4e-update-interval (* 10 60))
  (setq mu4e-get-mail-command "mbsync -a")
  ;;(setq mu4e-maildir "~/Mail")

  ;; Configure the function to use for sending mail
(setq message-send-mail-function 'smtpmail-send-it)

  (setq mu4e-contexts
        (list
         ;; Work account
         (make-mu4e-context
          :name "Work"
          :match-func
            (lambda (msg)
              (when msg
                (string-prefix-p "/nepgroup" (mu4e-message-field msg :maildir))))
          :vars '((user-mail-address . "kkruse@nepgroup.com")
                  (user-full-name    . "Kim Kruse")
		  (smtpmail-smtp-server  . "smtp.office365.com")
                  (smtpmail-smtp-service . 587)
                  (smtpmail-stream-type  . ssl)
                  (mu4e-drafts-folder  . "/nepgroup/Drafts")
                  (mu4e-sent-folder  . "/nepgroup/Sendt Post")
                  (mu4e-refile-folder  . "/nepgroup/Indbakke")
                  (mu4e-trash-folder  . "/nepgroup/Slettet Post")))

	 ;; Hotmail account
;;         (make-mu4e-context
;;          :name "Hotmail"
;;          :match-func
;;            (lambda (msg)
;;              (when msg
;;                (string-prefix-p "/Hotmail" (mu4e-message-field msg :maildir))))
;;          :vars '((user-mail-address . "kimkruse@hotmail.com")
;;                  (user-full-name    . "Kim Kruse")
;;		  (smtpmail-smtp-server  . "smtp-mail.outlook.com")
;;                  (smtpmail-smtp-service . 58;;7)
;;                  (smtpmail-stream-type  . st;;arttls)
;;                  (mu4e-drafts-folder  . "/Ho;;tmail/Drafts")
;;                  (mu4e-sent-folder  . "/Hotm;;ail/Sent Mail")
;;                  (mu4e-refile-folder  . "/Ho;;tmail/Alle mails")
;;                  (mu4e-trash-folder  . "/Hot;;mail/Trash")))

         ;; Personal account
         (make-mu4e-context
          :name "Personal"
          :match-func
            (lambda (msg)
              (when msg
                (string-prefix-p "/kimkruse" (mu4e-message-field msg :maildir))))
          :vars '((user-mail-address . "kim@kimkruse.xyz")
                  (user-full-name    . "Kim Kruse")
		  (smtpmail-smtp-server . "mail.kimkruse.xyz")
		  (smtpmail-smtp-service . 587)
		  (smtpmail-stream-type . starttls)
                  (mu4e-drafts-folder  . "/kimkruse/Drafts")
                  (mu4e-sent-folder  . "/kimkruse/Sent")
                  (mu4e-refile-folder  . "/kimkruse/Archive")
                  (mu4e-trash-folder  . "/kimkruse/Trash")))))

  (setq mu4e-maildir-shortcuts
      '(("/kimkruse/Inbox"             . ?i)
        ("/kimkruse/Sent" . ?s)
        ("/nepgroup/INBOX"    . ?d)
        ("/nepgroup/'Sendt Post'"     . ?t))))
#+end_src
* Custom.el
By default, Rational Emacs keeps your config file clean. All the customization
settings that Emacs normally automatically adds to your config.el go into
the file =custom.el= instead. If you don't want this, set the respective
variable to =nil=:
#+begin_src emacs-lisp
  ;; To not load `custom.el' after `config.el', uncomment this line.
  ;; (setq rational-load-custom-file nil)
#+end_src

* Tangling to early-config.el
If you need to make settings to =early-config.el=, you can do that from here, too.
Just begin the source code block with:
#+begin_src org
  ,#+begin_src emacs-lisp :tangle ./early-config.el
#+end_src


emacs/history
;; -*- mode: emacs-lisp; coding: utf-8-unix -*-
;; Minibuffer history file, automatically generated by ‘savehist’.

(setq savehist-minibuffer-history-variables '(ivy-history empty-history counsel-M-x-history buffer-name-history file-name-history))
(setq ivy-history '(#("perspective-exwm" 0 16 (ivy-index 0)) #("vterm" 0 5 (ivy-index 1)) #("*vterm*" 0 7 (ivy-index 1)) #("exwm" 0 4 (ivy-index 0))))
(setq counsel-M-x-history '(#("^vterm" 0 6 (ivy-index 2)) "vterm" #("package-install" 0 15 (ivy-index 0)) #("^all-t" 0 6 (ivy-index 1)) "all-the-icons-install-fonts" "vterm" #("^package-install" 0 16 (ivy-index 0)) "vterm" #("^mu4e" 0 5 (ivy-index 121)) "mu4e" #("^server-start" 0 13 (ivy-index 0)) "server-start" "vterm" "vterm" "vterm" "vterm" #("^all-the-ico" 0 12 (ivy-index 7)) "all-the-icons-install-fonts"))
(setq buffer-name-history '(#(".xprofile" 0 9 (ivy-index 0)) #("EXWM.desktop" 0 12 (ivy-index 0)) #("default.conf" 0 12 (ivy-index 0)) #("Qogir-dark#.kvconfig" 0 20 (ivy-index 0)) #("kvantum.kvconfig" 0 16 (ivy-index 0)) #("authinfo" 0 8 (ivy-index 0)) #(".authinfo.gpg" 0 13 (ivy-index 0)) #(".authinfo" 0 9 (ivy-index 0)) #("sxhkdrc" 0 7 (ivy-index 0)) #("kim" 0 3 (ivy-index 0)) #("battery.el" 0 10 (ivy-index 0)) #("*Directory*" 0 11 (ivy-index 0))))
(setq file-name-history '(#("/home/kim/.xprofile" 0 19 (ivy-index 0)) #("/sudo:root@Arch:/usr/share/xsessions/EXWM.desktop" 0 49 (ivy-index 0)) #("/sudo:root@Arch:/usr/lib/sddm/sddm.conf.d/default.conf" 0 54 (ivy-index 0)) #("/home/kim/.ssh/id_rsa.pub" 0 25 (ivy-index 0)) #("/home/kim/.config/Kvantum/Qogir-dark#/Qogir-dark#.kvconfig" 0 58 (ivy-index 0)) #("/home/kim/.config/Kvantum/kvantum.kvconfig" 0 42 (ivy-index 0)) #("/home/kim/.config/leftwm/" 0 25 (ivy-index 0)) #("/home/kim/.authinfo.gpg" 0 23 (ivy-index 0)) #("/home/kim/authinfo" 0 18 (ivy-index 2)) #("/home/kim/.authinfo" 0 19 (ivy-index 0)) #("/home/kim/.config/sxhkd/" 0 24 (ivy-index 0)) #("/home/kim/.config/leftwm/config.toml" 0 36 (ivy-index 0)) #("/sudo:root@kim-20fas7a000:/usr/share/xsessions/EXWM.desktop" 0 59 (ivy-index 0)) #("/home/kim/.dots/emacs/" 0 22 (ivy-index 0)) #("/home/kim/.config/emacs/myinit.org" 0 34 (ivy-index 1))))


.rational-emacs/config.el
;;; config.el -- Example Rational Emacs user customization file -*- lexical-binding: t; -*-
;; This file is generated from config.org. If you want to edit the
;; configuration, DO NOT edit config.el, edit config.org, instead.

;; Tangle the code blocks to config.el on save.
(defun org-babel-tangle-config ()
  (when (string-equal (buffer-file-name)
                      (expand-file-name "config.org" rational-config-path))
    ;; Dynamic scoping to the rescue
    (let ((org-confirm-babel-evaluate nil))
      (org-babel-tangle))))

(add-hook 'org-mode-hook
          (lambda ()
            (add-hook 'after-save-hook #'org-babel-tangle-config)))

(require 'rational-defaults)    ; Sensible default settings for Emacs
(require 'rational-use-package) ; Configuration for `use-package`
(require 'rational-updates)     ; Tools to upgrade Rational Emacs
(require 'rational-completion)  ; selection framework based on `vertico`
(require 'rational-ui)          ; Better UI experience (modeline etc.)
(require 'rational-windows)     ; Window management configuration
(require 'rational-editing)     ; Whitspace trimming, auto parens etc.
;(require 'rational-evil)        ; An `evil-mode` configuration
(require 'rational-org)         ; org-appear, clickable hyperlinks etc.
(require 'rational-project)     ; built-in alternative to projectile
(require 'rational-speedbar)    ; built-in file-tree
(require 'rational-screencast)  ; show current command and binding in modeline
(require 'rational-compile)     ; automatically compile some emacs lisp files

(custom-set-variables
   '(rational-ui-default-font
     '(:font "JetBrains Mono" :weight light :height 185)))

(rational-package-install-package 'doom-themes)
(load-theme 'doom-wilmersdorf t)

(customize-set-variable 'rational-startup-inhibit-splash t)

(use-package doom-modeline
   :ensure t
   :init (doom-modeline-mode 1)
   :custom ((doom-modeline-height 35)
            (doom-modeline-bar-width 10)
            (doom-modeline-buffer-file-name-style 'auto)
            (doom-modeline-icon (display-graphic-p))
            (doom-modeline-major-mode-icon t)
            (doom-modeline-major-mode-color-icon t)
            (doom-modeline-enable-word-count t)
            (doom-modeline-modal-icon t)
            (doom-modeline-mu4e nil)
            (mu4e-alert-enable-mode-line-display)))

 (display-battery-mode 1)

 ;; Turn on `display-time-mode' if you don't use an external bar.
 (setq display-time-default-load-average nil)

 (display-time-mode t)

 (setq display-time-24hr-format 1)

 ;;(display-time-mode nil)
 ;;(setq display-time-24hr-format nil)

(setenv "GPG_AGENT_INFO" nil)  ;; use emacs pinentry
    (setq auth-source-debug t)

    (setq epg-gpg-program "gpg2")  ;; not necessary
    (require 'epa-file)
    (epa-file-enable)
;;(setq epa-pinentry-mode 'loopback)
(setq epg-pinentry-mode 'loopback)

    (require 'org-crypt)
    (org-crypt-use-before-save-magic)

(defun efs/lookup-password (&rest keys)
  (let ((result (apply #'auth-source-search keys)))
    (if result
	(funcall (plist-get (car result) :secret))
      nil)))

(use-package counsel
    :ensure t)

  (use-package ivy
    :ensure t
    :diminish (ivy-mode)
    :bind (("C-x b" . ivy-switch-buffer))
    :config
    (ivy-mode 1)
    (setq ivy-use-virtual-buffers t)
    (setq ivy-display-style 'fancy)
    (setq enable-recursive-minibuffers t)
  )
    (use-package swiper
      :ensure try
      :bind (("C-s" . swiper)
	     ("C-r" . swiper)
	     ("C-c C-r" . ivy-resume)
	     ("M-x" . counsel-M-x)
	     ("C-x C-f" . counsel-find-file))
      :config
      (progn
	(ivy-mode 1)
	(setq ivy-use-virtual-buffers t)
	(setq ivy-display-style 'fancy)
	(define-key read-expression-map (kbd "C-r") 'counsel-expression-history)
	))

  (use-package avy
    :ensure t
    :bind ("M-s" . avy-goto-char))

(use-package mu4e
  :ensure nil
  :config

  ;;this is set to 't' to avoid mail syncing issues when using mbsync
  (setq mu4e-change-filenames-when-moving t)

  ;; refresh mail using isync every 10 minutes
  (setq mu4e-update-interval (* 10 60))
  (setq mu4e-get-mail-command "mbsync -a")
  ;;(setq mu4e-maildir "~/Mail")

  ;; Configure the function to use for sending mail
(setq message-send-mail-function 'smtpmail-send-it)

  (setq mu4e-contexts
        (list
         ;; Work account
         (make-mu4e-context
          :name "Work"
          :match-func
            (lambda (msg)
              (when msg
                (string-prefix-p "/nepgroup" (mu4e-message-field msg :maildir))))
          :vars '((user-mail-address . "kkruse@nepgroup.com")
                  (user-full-name    . "Kim Kruse")
		  (smtpmail-smtp-server  . "smtp.office365.com")
                  (smtpmail-smtp-service . 587)
                  (smtpmail-stream-type  . ssl)
                  (mu4e-drafts-folder  . "/nepgroup/Drafts")
                  (mu4e-sent-folder  . "/nepgroup/Sendt Post")
                  (mu4e-refile-folder  . "/nepgroup/Indbakke")
                  (mu4e-trash-folder  . "/nepgroup/Slettet Post")))

	 ;; Hotmail account
;;         (make-mu4e-context
;;          :name "Hotmail"
;;          :match-func
;;            (lambda (msg)
;;              (when msg
;;                (string-prefix-p "/Hotmail" (mu4e-message-field msg :maildir))))
;;          :vars '((user-mail-address . "kimkruse@hotmail.com")
;;                  (user-full-name    . "Kim Kruse")
;;		  (smtpmail-smtp-server  . "smtp-mail.outlook.com")
;;                  (smtpmail-smtp-service . 58;;7)
;;                  (smtpmail-stream-type  . st;;arttls)
;;                  (mu4e-drafts-folder  . "/Ho;;tmail/Drafts")
;;                  (mu4e-sent-folder  . "/Hotm;;ail/Sent Mail")
;;                  (mu4e-refile-folder  . "/Ho;;tmail/Alle mails")
;;                  (mu4e-trash-folder  . "/Hot;;mail/Trash")))

         ;; Personal account
         (make-mu4e-context
          :name "Personal"
          :match-func
            (lambda (msg)
              (when msg
                (string-prefix-p "/kimkruse" (mu4e-message-field msg :maildir))))
          :vars '((user-mail-address . "kim@kimkruse.xyz")
                  (user-full-name    . "Kim Kruse")
		  (smtpmail-smtp-server . "mail.kimkruse.xyz")
		  (smtpmail-smtp-service . 587)
		  (smtpmail-stream-type . starttls)
                  (mu4e-drafts-folder  . "/kimkruse/Drafts")
                  (mu4e-sent-folder  . "/kimkruse/Sent")
                  (mu4e-refile-folder  . "/kimkruse/Archive")
                  (mu4e-trash-folder  . "/kimkruse/Trash")))))

  (setq mu4e-maildir-shortcuts
      '(("/kimkruse/Inbox"             . ?i)
        ("/kimkruse/Sent" . ?s)
        ("/nepgroup/INBOX"    . ?d)
        ("/nepgroup/'Sendt Post'"     . ?t))))

;; To not load `custom.el' after `config.el', uncomment this line.
;; (setq rational-load-custom-file nil)
